### PR TITLE
Feature/dockerfile

### PR DESCRIPTION
**Pull request by radrow**
- added docker stuff
- more idiomatic app design

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -28,7 +30,21 @@ Launch the REPL
...
+For a consistent setup, a docker image can be created:
+
+```
+make docker
+```
+
+Then to start the dockerized REPL:
+
+```
+docker run -i aeternity/aerepl
```
Comment (dincho) [inline README.md:44]: make docker produces `aeternity/aerepl:local` so you'd probably want that :local tag in this cmd as well

Comment (radrow) [reply README.md:44]: right

File: src/aere_parse.erl [hunk 2]
```diff
@@ -3,10 +3,9 @@
 -export([ parse/1, words/1, commands/0, resolve_command/1 ]).
 
 -type parse_result() :: {atom(), string() | [string()]}
-                      | no_return()
-                      | skip.
```
Comment (ghallak) [inline src/aere_parse.erl:7]: @radrow Why have you decided to remove `skip` from the `parse_result` type even though the `parse` function returns `{skip, []}`?

File: src/aere_theme.erl [hunk 3]
```diff
@@ -150,7 +150,7 @@ render(Theme, ThemedTxt = {themed, _, _}) ->
     render(Theme, [ThemedTxt]);
 render(Theme, ThemedTxts) when is_list(ThemedTxts) ->
     AnsiStr = lists:flatten(lists:map(fun(T) -> apply_theme(Theme, T) end, ThemedTxts)),
-    string:trim(AnsiStr, both, unicode_util:whitespace()).
+    string:trim(AnsiStr, trailing, unicode_util:whitespace()).
```
Comment (ghallak) [inline src/aere_theme.erl:153]: What's the problem with trimming both sides? Indention or something else?

File: src/aere_repl.erl [hunk 4]
```diff
@@ -477,6 +483,11 @@ get_ready_chain(_) ->
     throw({repl_error, aere_msg:chain_not_ready()}).
 
 set_option(Option, Args, S = #repl_state{options = Opts}) ->
+    Locked = maps:get(locked_opts, Opts, []),
+    case lists:member(Option, [locked_opts|Locked]) of
+        true -> throw({repl_error, aere_msg:locked_option()});
+        _ -> ok
+    end,
```
Comment (ghallak) [inline src/aere_repl.erl:490]: This may look nicer:

```suggestion
    LockedOption = lists:member(Option, [locked_opts|Locked]),
    LockedOption andalso throw({repl_error, aere_msg:locked_option()}),
```

File: src/aere_repl.hrl [hunk 5]
```diff
@@ -7,6 +7,7 @@
         , call_value   := non_neg_integer()
         , print_format := print_format()
         , print_unit   := boolean()
+        , locked_opts  => [atom()]
```
Comment (ghallak) [inline src/aere_repl.hrl:10]: Why are you using an optional association here? Why not `locked_opts  := [atom()]`?

File: VERSION [hunk 6]
```diff
@@ -1 +1 @@
-2.0.0
+6.5.2+5.da93ca823
```
Comment (ghallak) [inline VERSION:1]: Why is this file needed? Where is it used? and why does it have the node version instead of the aerepl version?

File: rebar.config [hunk 7]
```diff
@@ -20,12 +22,12 @@
        ]}.
 
 {relx, [{release, {aerepl, "2.0.0"},
-         [aerepl, {aechannel, load}, {aecore, load}, {aeutils, load}]},
-        {lib_dirs, ["node/_build/dev1/rel/aeternity/lib"]},
+         [aerepl, {aechannel, load}, {aecore, load}, {aeutils, load}, {mnesia, load}, sasl]},
+        {lib_dirs, ["node/_build/dev1/rel/aeternity/lib", "_build/prod/lib/"]},
```
Comment (ghallak) [inline rebar.config:26]: Why do you need to load`mnesia` and include `_build/prod/lib`?

File: rebar.config [hunk 8]
```diff
@@ -20,12 +22,12 @@
...
         {dev_mode, true},
         {vm_args, "config/vm.args"},
         {sys_config, "config/sys.config"},
-        {include_erts, false},
+        {include_erts, true},
```
Comment (ghallak) [inline rebar.config:30]: Why do you need to include erts?

File: rebar.config (hunk 7 above)
Comment (radrow) [reply rebar.config:26]: There were some errors in the release. Didn't affect semantics as far as I checked, but I prefer clean run just in case

File: src/aere_repl_state.erl [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    S#repl_state.included_code .
+-spec included_code(aeso_syntax:ast(), state()) -> state().
+included_code(X, S) ->
+    S#repl_state{included_code = X}.
+-spec query_nonce(state()) -> non_neg_integer().
+query_nonce(S) ->
+    S#repl_state.query_nonce.
+-spec query_nonce(non_neg_integer(), state()) -> state().
+query_nonce(X, S) ->
+    S#repl_state{query_nonce = X}.
```
Comment (ghallak) [inline src/aere_repl_state.erl:99]: Just throw new lines between functions. This makes my eyes bleed.

File: src/aere_parse.erl (hunk 2 above)
Comment (radrow) [reply src/aere_parse.erl:7]: Because skip /= {skip, []}

File: src/aere_theme.erl (hunk 3 above)
Comment (radrow) [reply src/aere_theme.erl:153]: Yeah, indentation gets problematic in the first line. It's hard to handle when coloring characters are involved.

File: src/aere_repl.erl [hunk 10]
```diff
@@ -477,6 +483,11 @@ get_ready_chain(_) ->
     throw({repl_error, aere_msg:chain_not_ready()}).
 
 set_option(Option, Args, S = #repl_state{options = Opts}) ->
+    Locked = maps:get(locked_opts, Opts, []),
+    case lists:member(Option, [locked_opts|Locked]) of
```
[the remark this answers is not in the document]
Comment (radrow) [reply src/aere_repl.erl:487]: Because I don't want to let the user unlock a locked option (and give a meaningful message)

File: src/aere_repl.hrl (hunk 5 above)
Comment (radrow) [reply src/aere_repl.hrl:10]: Because it has a natural default which is an empty list

File: VERSION (hunk 6 above)
Comment (radrow) [reply VERSION:1]: I think you can remove it

File: rebar.config (hunk 8 above)
Comment (radrow) [reply rebar.config:30]: I don't remember, it was a fix to something but I forgot

File: src/aere_parse.erl (hunk 2 above)
Comment (ghallak) [reply src/aere_parse.erl:7]: Yeah, now I see that `{atom(), string() | [string()]}` fits `{skip, []}` as well. I missed that one.

File: VERSION (hunk 6 above)
Comment (ghallak) [reply VERSION]: Removed here 550c751bf4be84cb762c47299f174840a0c1299c

File: rebar.config [hunk 11]
```diff
@@ -9,6 +9,8 @@
                            {ref,"eb68fe3"}}},
         {aebytecode, {git, "https://github.com/aeternity/aebytecode.git",
                       {tag, "v3.1.1"}}},
+        {app_ctrl, {git, "https://github.com/aeternity/app_ctrl.git",
```
[the remark this answers is not in the document]
Comment (radrow) [reply rebar.config:12]: Release was complaining on lack of it

File: link_apps.bash [hunk 12]
```diff
@@ -0,0 +1,22 @@
+link_node_app () {
+    if ! [ -d $REBAR_BUILD_DIR/lib/$1 ]
+    then
+       ln -sdf $PWD/node/_build/dev1/lib/$1 $REBAR_BUILD_DIR/lib/$1;
```
Comment (ghallak) [inline link_apps.bash:4]: @radrow using `ln -d` is not possible on OS X because the `ln` command does not have a `-d` option.

As I understand from this [StackExchange comment](https://unix.stackexchange.com/a/267790) this option does not work on all filesystems.

File: src/aere_repl.erl (hunk 10 above)
[the remark this answers is not in the document]
Comment (ghallak) [reply src/aere_repl.erl]: I don't think this is really necessary since it's already included in the release version is being used for the docker image, so I have changed this back to false. If we found that this is causing any problems in the future, we can bring it back.

File: rebar.config (hunk 8 above)
Comment (ghallak) [reply rebar.config]: I don't think this is really necessary, since erts is already included in the release version that is being used for the docker image. I have changed this back to false. If we find that this is causing any problems in the future, we can bring it back.

File: rebar.config (hunk 11 above)
[the remark this answers is not in the document]
Comment (ghallak) [reply rebar.config:12]: It is not complaining about it for me. What command are you running to get that?

[the remark this answers is not in the document]
Comment (radrow) [reply rebar.config:12]: Just run the release from _build/prod/rel (or something like that) and you should see undef errors
